### DISCUSSION TITLE
Official taps vs core tap?

### DISCUSSION THREAD

**Discussion by tddschn** [Homebrew/discussions#268] [Everyday usage] [answered]
For formulas like `terraform`, `consul` and `clojure`, the offical taps (`hashicorp/tap`, clojure/tools`) are preferred over the core tap for the following reasons:
- recommended by the vendors
- more up-to-date
- vendor-controlled

And IMO, maintaining these formulas in the core tap has the following drawbacks:
- discourages users from using the official tap
- more work to do for the maintainers
- sometimes can be confusing for users

My questions:
- Why are these formulas included in the core tap (`Homebrew/core`)?
- Is there a way to set preference of the formula file used when installing a specific formula? (e.g. prefer `vendor/tap/tool` over `homebrew/core/tool` so that `tool` always refers to `vendor/tap/tool` when using homebrew by default unless specified otherwise on the command line.)

**Comment by carlocab** [comment]
Here’s a related discussion on Homebrew’s (deprecated) discourse: https://discourse.brew.sh/t/maintenance-of-formulas-by-vendor/7649

**Comment by SMillerDev** [comment]
> Why are these formulas included in the core tap (Homebrew/core)?

Because they are open source, people submitted and use them. And some software in core in turn depends on it.

> Is there a way to set preference of the formula file used when installing a specific formula? (e.g. prefer vendor/tap/tool over homebrew/core/tool so that tool always refers to vendor/tap/tool when using homebrew by default unless specified otherwise on the command line.)

No, because homebrew-core as a policy only tests with it's own formulae so we have no way to know if this works. Therefore it's unsupported and not likely to be added/accepted. If you want to change the dependencies of a formulae, you'll have to `brew extract` it and maintain it yourself.

**Reply by carlocab** [reply]
Broken formulae are also much harder to fix if they're not on Homebrew. Note that the CI infrastructure tests all of a formula's recursive dependents whenever that formula is updated. Testing-upon-update is also harder to implement if the formula that is being updated is outside of homebrew-core.

**Comment by kennedy** [comment]
@radeksimko

Has there been a decision on whether hashicorp will have their own tap?

**Reply by carlocab** [reply]
This might interest you: https://github.com/hashicorp/homebrew-tap

**Comment by radeksimko** [comment, marked as answer]
> And IMO, maintaining these formulas in the core tap has the following drawbacks:

I can't speak for any end users or core maintainers perception of this but we aimed to address the mentioned drawbacks in the main Readme of our mentioned tap: https://github.com/hashicorp/homebrew-tap#why-should-i-install-packages-from-this-tap

> - discourages users from using the official tap
> - more work to do for the maintainers

We _don't explicitly discourage_ anyone from installing from the core tap but AFAIK we would likely ask them to use _our_ tap if they ask for _our_ (typically paid) help. That I think is reasonable and fair. In the same way as Homebrew maintainers likely would not want to be debugging our own builds, we don't want to be debugging theirs. I don't expect we'd refuse to support customers that install from the core tap but it typically helps to "remove a variable from the equation" when debugging.

> - sometimes can be confusing for users

To reduce some confusion we use the `conflicts_with` stanzas (as discussed in the linked Discourse).

I expect there will still be users who can basically say _"I installed this vendor's software so if anything doesn't work it's their fault"_ and that is only something we can correct in more direct support conversations I mentioned and upfront in the docs I linked above. I don't see what more we can do other than keep correcting that false assumption and their understanding of the reality of OSS and packaging. 🤷🏻‍♂️ 

On the technical side of things there were some ideas floated [here](https://discourse.brew.sh/t/maintenance-of-formulas-by-vendor/7649/14) on improving the Homebrew UX which could further reduce possible confusion but personally I haven't had the capacity to implement any of that.

**Reply by Propfend** [reply]
is there some reason why the core package is not as up to date as the official tap? I saw there are some softwares which can optmize some pipeline to create GH taps, does one have some easy way to do so with core packages in homebrew or they need to be created manually?

**Reply by gromgit** [reply]
> is there some reason why the core package is not as up to date as the official tap?

In general, it's almost always one of the following:
1. The formula was bumped, but Homebrew CI builds are failing. Official taps can always build by hand and package manually for their own releases.
2. The new version was not tagged for release. Pre-releases in particular are ignored by Homebrew's livecheck mechanism, but may be bumped anyway on the official tap.
3. The formula is no longer being livechecked by Homebrew for various reasons (e.g. `vault` has been disabled in Homebrew core because its current license no longer meets Homebrew's guidelines).

**Reply by Propfend** [reply]
thats interesting, but then you talk like the Homebrew will add it automatically to core from your tap, is it true?

**Reply by gromgit** [reply]
> you talk like the Homebrew will add it automatically to core from your tap

I'm not sure how you concluded that from what I wrote.

Homebrew formulae are always _created_ by human beings who want to add the software they're interested in. If no one's interested enough in project `X` to write the initial formula for it, Homebrew will never host that software.

After they've been added, _and_ if there's a _livecheck_ block in a formula that specifies how to check for new releases, _then_ Homebrew CI will auto-bump that formula when it detects a new release. At that point, my previous answer to your question applies.
